Longhorn Project Security Self-Assessment - Security Pals (#1183)
cncf/tag-security: assessments/projects/longhorn/self-assessment.md

> # Longhorn Self-assessment
> This is the Longhorn's Self-assessment documentation that we wrote after we began thinking about the security of the project, determining gaps in its security, and preparing security documentation for Longhorn's users. This project is currently in the CNCF Incubation phase preparing for Graduation.
> 
> 
> ## Table of contents

Reviewer: ```suggestion
# Longhorn Self-assessment

This assessment was created by community members as part of the [Security Pals](https://github.com/cncf/tag-security/issues/1102) process, and is currently pending changes from the maintainer team.

The Longhorn self-assessment documentation was written with security in mind. It was written after carefully considering the security of the project, determining gaps in its security, and preparing security documentation for Longhorn's users. This project is currently in the CNCF Incubation phase, preparing for Graduation.

## Table of contents
```
@ragashreeshekar

* [Metadata](#metadata)
  * [Security links](#security-links)
* [Overview](#overview)
  * [Actors](#actors)
  * [Actions](#actions)
  * [Background](#background)
  * [Goals](#goals)
  * [Non-goals](#non-goals)
* [Self-assessment use](#self-assessment-use)
* [Threat Modelling](#Threat-Modelling)
* [Security functions and features](#security-functions-and-features)
* [Project compliance](#project-compliance)
* [Secure development practices](#secure-development-practices)
* [Security issue resolution](#security-issue-resolution)
* [Appendix](#appendix)
* [Action items](#Action-Items)

## Metadata

> |   |  |
> | -- | -- |
> | Software | [A link to the Longhorn’s repository.](https://github.com/longhorn/longhorn) |
> | Security Provider | No |
> | Languages | Python, Shell, Mustache |
> | SBOM | Longhorn does not currently generate an SBOM on release |

Reviewer: ```suggestion
|   |  |
| -- | -- |
| Assessment Stage | Incomplete |
| Software | [A link to the Longhorn’s repository.](https://github.com/longhorn/longhorn) |
| Security Provider | No |
| Languages | Python, Shell, Mustache |
| SBOM | Longhorn does not currently generate an SBOM on release |
```
@ragashreeshekar

* Data Analytics and Processing: Longhorn is not designed for direct involvement in data analytics or processing tasks. Its role is confined to the storage and retrieval of data for such operations, not processing or analysing the data itself.
* Limitless Data Storage: While Longhorn facilitates data storage, it is not designed to support unlimited data storage or handle scenarios where the volume of data could potentially overwhelm the storage capacity or incur excessive costs.
* Network Infrastructure Management: Longhorn's scope does not include the management or configuration of underlying network infrastructure. It focuses on the storage aspect within the given network environment.
* Comprehensive Data Security: While Longhorn implements measures to secure data, it does not offer a complete data security solution. Users are responsible for implementing additional security measures according to their specific requirements, such as network security or end-to-end encryption of sensitive data.

## Self-assessment use

> This self-assessment is created by the Longhorn team to perform an internal analysis of the project's security. It is not intended to provide a security audit of Longhorn, or function as an independent assessment or attestation of Longhorn's security health.

Reviewer 2: ```suggestion
This self-assessment is not intended to provide a security audit of Longhorn, or function as an independent assessment or attestation of Longhorn's security health.
```

- Restrict access to internal metadata and configuration details.

#### 3.5 Denial of Service (DoS)
**Threats**:
- Overloading the Longhorn system, leading to unavailability.
- Exploiting vulnerabilities to crash the system or degrade performance.

**Mitigations**:
- Implement rate limiting and access controls.
- Design for high availability and resilience.
- Regularly update and patch to address known vulnerabilities.

#### 3.6 Elevation of Privilege
**Threats**:
- Exploitation of vulnerabilities to gain higher privileges.
- Unauthorised access leading to control over Longhorn operations.

**Mitigations**:
- Adhere to the principle of least privilege in access controls.
- Regular security assessments and penetration testing.
- Monitor and promptly update software components to address vulnerabilities.

### 4. Conclusion
This STRIDE threat model for Longhorn identifies key areas of potential security risks and provides a foundation for implementing effective security measures. Regular updates, vigilant monitoring, and adherence to security best practices are essential to maintain the security and integrity of the Longhorn system.

</details>

<details>
 

> <summary>Longhorn Lightweight Threat Model</summary>

Reviewer 3: @ragashreeshekar Is this supposed to be inlined here or moved to a different document?

Reviewer 3: It is also mentioned in the document a few places.  This may be fine but would need to be patched up if it is externalized.

Reviewer 2: Updated the document and its references

## Table of contents

* [Metadata](#metadata)
  * [Security links](#security-links)
* [Overview](#overview)
  * [Actors](#actors)
  * [Actions](#actions)
  * [Background](#background)
  * [Goals](#goals)
  * [Non-goals](#non-goals)
* [Self-assessment use](#self-assessment-use)
* [Threat Modelling](#Threat-Modelling)
* [Security functions and features](#security-functions-and-features)
* [Project compliance](#project-compliance)
* [Secure development practices](#secure-development-practices)
* [Security issue resolution](#security-issue-resolution)
* [Appendix](#appendix)
* [Action items](#Action-Items)

## Metadata

|   |  |
| -- | -- |
| Software | [A link to the Longhorn’s repository.](https://github.com/longhorn/longhorn) |
| Security Provider | No |
| Languages | Python, Shell, Mustache |
| SBOM | The information is not readily available online and we have requested for the same. This is an action item for our subsequent week |
| | |

### Security links

| Doc | url |
| -- | -- |

> | Security file | https://longhorn.io/docs/1.5.3/advanced-resources/security/ |

Reviewer 2: There is no content in this page. Can you please provide references to the secure development practices or other security content stated in this assessment?

Author: Hello, Ms. Ragashree Shekar. Thank you for the question. 
The URL points to the landing page of Security of Longhorn. While this page does not have any major information, it serves as an introduction/landing for security, and it is within the subpages that we can find more information on the Security.
The links to those sub-pages are also provided in the subsequent two rows of the table here. Hope that helps :) Thanks!

Author: If there are any questions, I will be happy to answer :) If there are none, may I go ahead and resole this conversation?

Author: As we are nearing the deadline, we are trying our best to not leave any unresolved questions and requests. Since there are no follow up questions for this thread, I am resolving this conversation. Please prompt us with any questions if you have any in the future and we will be happy to get back to you as soon as possible. Thanks!

| Storage                        | Data loss or corruption                      | Replication across nodes, regular snapshots, and backups               | [Longhorn Documentation](https://longhorn.io/docs/) |
| Authentication                 | Unauthorized system access                   | Strong authentication mechanisms, integration with Kubernetes auth      | [Kubernetes Authentication](https://kubernetes.io/docs/reference/access-authn-authz/authentication/) |
| Authorization (Access Control) | Unauthorized actions within the storage system| Strict RBAC policies, least privilege access                           | [Kubernetes Authorization](https://kubernetes.io/docs/reference/access-authn-authz/authorization/) |
| Audit and Logging              | Lack of incident visibility or forensics     | Enable detailed logging, integrate with monitoring tools               | [Kubernetes Logging Architecture](https://kubernetes.io/docs/concepts/cluster-administration/logging/) |

> ### Recommendations

Reviewer 2: This section is meant to provide recommendations to the project. Does this include all your recommendation to the project to enhance their security posture? Some of these look like recommendations for the end user and may not be relevant in this context.

Author: Hello, Ms. Ragashree Shekar. Thank you for the feedback and the question. The recommendations in this section were targeted for the project itself, aimed at enhancing its overall security posture, not for the end users. But I think some additional explanations or clarifications could help. Thank you for the feedback, we will add some explanations in this section.

Reviewer 4: Following discussions from the review team, we are asking that Security Pals please remove any recommendations from the scope of this document. You may optionally communicate recommendations for projects in a more accessible location, such as in a project community channel or GitHub Issue.

Author: Hello, Mr. Knight. Thank you for the comment. We have indeed also communicated our feedback and are trying to discuss the recommendations and finalise the action items with the maintainers of Longhorn. In the meantime, as per Mr. Pranava's instructions, we have added the Action Items and Recommendations under threat modelling to illustrate the future steps towards the security analysis. Perhaps, would it help to address these points within an appendix section at the bottom, Mr. Knight? 
Do let us know what you think, thanks in advance!

- Regularly update and patch Longhorn and Kubernetes components.
- Ensure that all communication is encrypted, both in transit and at rest.
- Conduct periodic access reviews and privilege audits.

### Conclusion

- Critical findings should be promptly disclosed to the community.
- Consideration for additional threat modelling tools, such as attack trees or matrices, to further analyse complex threats.
</details>

The attact tree for a generic high-level threat scenario is depicted:

<p align="center"><img alt="image" src="https://github.com/Makesh-Srinivasan/ISP_A3/assets/66047630/4f88f24c-14e9-40c1-92e5-ea3761806556"></p>

Each branch of this tree represents a step or method an attacker might use to progress their attack. The end goal is unauthorized access to and exfiltration of sensitive data stored within Longhorn. This attack tree can be used to identify potential vulnerabilities, assess risks, and develop strategies to mitigate or prevent such attacks.

<hr>

> ## Security functions and features

Reviewer 2: These features are meant to aid threat modelling. Hope the threats and recommendations consider them.

Author: Dear Ms. Ragashree Shekar, thank you for the comment. Yes, we did indeed use these features in understanding and later building our threat models and analyse the possible threats. Best regards!

* Kubernetes' Role-Based Access Control (RBAC) Integration: By leveraging Kubernetes RBAC, Longhorn ensures that only authorised users and processes can manage volumes, snapshots, and backups, thus enforcing the principle of least privilege.
* Health Checks and Monitoring: Longhorn's continuous health checks and monitoring of volumes and replicas are crucial for early detection of potential issues, allowing for proactive remediation before they escalate into security incidents.
* Backup Data Validation: Before restoring data from backups, Longhorn validates the backup integrity, ensuring that the data has not been tampered with or corrupted, which is essential for the reliability of the restoration process.

## Project compliance
* Compliance: Longhorn adheres to best practices for Kubernetes storage solutions, ensuring it aligns with the security and operational standards expected within the cloud-native ecosystem. However, specific compliance certifications or adherence to standards like PCI-DSS, COBIT, ISO, or GDPR have not been documented.

## Secure development practices

Development Pipeline

> * Version Control and Commit Signing: Longhorn uses Git for version control. Contributors are encouraged to sign commits to verify the identity of contributors. More details can be found in the [Getting Started with Longhorn Development](https://github.com/longhorn/longhorn/wiki/Getting-started-with-Longhorn-Development) guide.

Reviewer 2: I couldn't find references to commit signing in this document. Could you please point me to the commit signing or section with secure development practices in this document?

Author: Dear Ms. Ragashree Shekar, thank you for this question. An updated link is now added to the self assessment detailing the guidelines on contributing and signing commits. Hope that helps! Thank you :)

> * Code Reviews: Each pull request requires a thorough review by at least two maintainers to ensure quality and security.

Reviewer 2: In some of the closed PRs, this doesn't seem to be the case. Please revisit and provide the reference to this process.

Reviewer 5: Thank you for pointing that out, we have fix it.

* Post-Incident Analysis: After resolving the incident, the team conducts a post-mortem to understand the root cause, improve future response efforts, and enhance security measures to prevent similar incidents.

## Appendix
Known Issues Over Time
* Longhorn tracks all issues publicly on their [GitHub issues page](https://github.com/longhorn/longhorn/issues). Any security vulnerabilities discovered in the past are documented there along with their resolutions.
* The project maintains a [security advisory](https://github.com/longhorn/longhorn/security/advisories) page where they publish details about the security vulnerabilities found.
* Statistics on vulnerabilities found and fixed are not explicitly listed unless provided on the project’s repository or documentation. If available, data about the effectiveness of code reviews and automated testing in catching issues would be included.
* For a detailed history of known issues, visit: [Known Issues Over Time](https://longhorn.io/blog/cve-2021-36779-36780/).

CII Best Practices
* Longhorn is actively working towards meeting the CII Best Practices criteria.

> * Longhorn’s commitment to security can be observed in their adherence to these practices, such as using automated testing, code review, and maintaining a public version control repository. More details can be found here: [CII Best Practices](https://longhorn.io/docs/1.5.3/best-practices/)

Reviewer 6: The link provided talks about the best practices to use Longhorn, What needs to be documented here is the best practices followed by Long horn project in updating, maintaining and securing the project. If Longhorn doesn't have CII badge for the project or any corresponding entry, add that as an action item for the project.

Author: Thank you, Mr. Pranava. We have updated the action item to include acquiring a CII badge as well. Thank you for pointing this out

  - [JT4 LLC](https://www.jt4llc.com)
  - [DgPays](https://dgpays.com/en.html)
  - [Amdocs](http://amdocs.com/)

Related Projects / Vendors
* Longhorn is often compared to other CNCF projects or cloud-native storage solutions such as [Rook](https://github.com/rook/rook) and [OpenEBS](https://github.com/openebs/openebs).
* Especially Rook is often the most compared with Longhorn since both are Kubernetes-native. Rook is an open-source cloud-native storage utility for Kubernetes that aims to automate some of the tasks of a storage administrator, such as programmatic storage, migration, disaster recovery, monitoring, and resource management. Longhorn is valued to be easier to work on from scratch, but Rook is preferred when dealing with larger clusters in PRD.
* Prospective users are interested in differences in performance, scalability, ease of use, and specific features like snapshotting and backup/restore capabilities.
* For more on related projects and vendors, see: [Related Projects / Vendors](https://www.cncf.io/projects/longhorn/).

> ## Action Items

Reviewer 6: Your team has done a good work, but update the action item section with the all possible actions that you would like the project undertake in order make the project more secure and compliant. Populate the section with only the action items which someone can implement, you can keep a sub-section for open questions.

Author: Thank you for clarifying it, Mr. Pranava. We will update this section accordingly

Author: Hello, Mr. Knight. Thank you for the comment. We have indeed also communicated our feedback and are trying to discuss the recommendations and finalise the action items with the maintainers of Longhorn. In the meantime, as per Mr. Pranava's instructions, we have added the Action Items and Recommendations under threat modelling to illustrate the future steps towards the security analysis. Perhaps, would it help to address these points within an appendix section at the bottom, Mr. Knight? 
Do let us know what you think, thanks in advance!

Author: Thank you, Mr Knight. We have discussed and agree that it would be best to communicate the action items as well as the recommendations to the maintainers of Longhorn via their forums and chats instead of keeping them here in the security self assessment. We are trying our best to posit our ideas as well as clarify our questions with them. If any aspects change, we will address them as needed in the later documentation iterations. Thanks!

# Self-assessment
This is the Longhorn's Self-assessment documentation that we wrote after we began thinking about the security of the project, determining gaps in its security, and preparing security documentation for Longhorn's users. This project is currently in the CNCF Incubation phase preparing for Graduation.

> # Self-assessment outline

Reviewer 4: ```suggestion
```

## Table of contents

* [Metadata](#metadata)
  * [Security links](#security-links)
* [Overview](#overview)
  * [Actors](#actors)
  * [Actions](#actions)
  * [Background](#background)
  * [Goals](#goals)
  * [Non-goals](#non-goals)
* [Self-assessment use](#self-assessment-use)
* [Threat Modelling](#Threat-Modelling)
* [Security functions and features](#security-functions-and-features)
* [Project compliance](#project-compliance)
* [Secure development practices](#secure-development-practices)
* [Security issue resolution](#security-issue-resolution)
* [Appendix](#appendix)
* [Action items](#Action-Items)

## Metadata

> A table at the top for quick reference information, later used for indexing.

Reviewer 4: ```suggestion
```

Reviewer 4: Please check to ensure that no text from the template is copied over into this document

# Longhorn Self-assessment
This is the Longhorn's Self-assessment documentation that we wrote after we began thinking about the security of the project, determining gaps in its security, and preparing security documentation for Longhorn's users. This project is currently in the CNCF Incubation phase preparing for Graduation.

## Table of contents

* [Metadata](#metadata)
  * [Security links](#security-links)
* [Overview](#overview)
  * [Actors](#actors)
  * [Actions](#actions)
  * [Background](#background)
  * [Goals](#goals)
  * [Non-goals](#non-goals)
* [Self-assessment use](#self-assessment-use)
* [Threat Modelling](#Threat-Modelling)
* [Security functions and features](#security-functions-and-features)
* [Project compliance](#project-compliance)
* [Secure development practices](#secure-development-practices)
* [Security issue resolution](#security-issue-resolution)
* [Appendix](#appendix)
* [Action items](#Action-Items)

## Metadata

|   |  |
| -- | -- |
| Software | [A link to the Longhorn’s repository.](https://github.com/longhorn/longhorn) |
| Security Provider | No |
| Languages | Python, Shell, Mustache |

> | SBOM | [Longhorn_SBOM](https://github.com/Makesh-Srinivasan/ISP_A3/blob/360765d0b998c16fbea029c6964f96628f31a154/assessments/projects/longhorn/longhorn_SBOM.json) <br/> (generated by GitHub SBOM exporting tool - further document can be accessed: [GitHub SBOM export](https://docs.github.com/en/code-security/supply-chain-security/understanding-your-software-supply-chain/exporting-a-software-bill-of-materials-for-your-repository) ) |

Reviewer 4: ```suggestion
| SBOM | Longhorn does not currently generate an SBOM on release |
```
